File: tests/playwright/tests/objects/forms/auth/signup.form.ts
import { expect, Locator, test } from '@playwright/test';

export class SignUpForm {
  private container: Locator;
  private userNameInput: Locator;
  private userNameError: Locator;
  private emailInput: Locator;
  private emailError: Locator;
  private passwordInput: Locator;
  private passwordError: Locator;
  private signUpButton: Locator;
  private goSignUpButton: Locator;
  private goSignInButton: Locator;

  constructor(container: Locator) {
    this.container = container;
    this.userNameInput = container.getByTestId('input-username');
    this.userNameError = container.getByTestId('error-username');
    this.emailInput = container.getByTestId('input-email');
    this.emailError = container.getByTestId('error-email');
    this.passwordInput = container.getByTestId('input-password');
    this.passwordError = container.getByTestId('error-password');
    this.signUpButton = container.getByTestId('button-auth');
    this.goSignUpButton = container.getByTestId('button-go-signup');
    this.goSignInButton = container.getByTestId('button-go-signin');
  }

  async shouldBeVisible() {
    await test.step('"Sign up" form should be visible', async () => {
      await this.userNameInputShouldBeVisible();
      await this.emailInputShouldBeVisible();
      await this.passwordInputShouldBeVisible();
      await this.signUpButtonShouldBeVisible();
      await this.goSignInButtonShouldBeVisible();
    });
  }

  async fillForm(userName: string, email: string, password: string) {
    await test.step(`Fill "Sign up" form with "${userName}", "${email}", "${password}"`, async () => {
      await this.fillUserNameInput(userName);
      await this.fillEmailInput(email);
      await this.fillPasswordInput(password);
    });
  }

  async setForm(userName: string, email: string, password: string) {
    await test.step(`Set "Sign up" form to "${userName}", "${email}", "${password}"`, async () => {
      await this.setUserNameInput(userName);
      await this.setEmailInput(email);
      await this.setPasswordInput(password);
    });
  }

  async errorsShouldNotBeVisible() {
    await test.step('"Sign up" form errors should not be visible', async () => {
      await this.userNameErrorShouldNotBeVisible();
      await this.emailErrorShouldNotBeVisible();
      await this.passwordErrorShouldNotBeVisible();
    });
  }

  async userNameInputShouldBeVisible() {
    await test.step(`"Username" input should be visible`, async () => {
      await expect(this.userNameInput).toBeVisible();
    });
  }

  async fillUserNameInput(userName: string) {
    await test.step(`Fill "Username" input with "${userName}"`, async () => {
      await this.userNameInput.fill(userName);
    });
  }

  async setUserNameInput(userName: string) {
    await test.step(`Set "Username" input to "${userName}"`, async () => {
      await this.userNameInput.fill(userName);
      await this.userNameInput.blur();
    });
  }

  async userNameErrorShouldBeVisible(errorMessage?: string) {
    if (!errorMessage) {
      await test.step(`"Username" error should be visible`, async () => {
        await expect(this.userNameError).toBeVisible();
      });
    } else {
      await test.step(`"Username" error: "${errorMessage}"  should be visible`, async () => {
        await expect(this.userNameError).toBeVisible();
        await expect(this.userNameError).toHaveText(errorMessage);
      });
    }
  }

  async userNameErrorShouldNotBeVisible() {
    await test.step(`"Username" error should not be visible`, async () => {
      await expect(this.userNameError).not.toBeVisible();
    });
  }

  async emailInputShouldBeVisible() {
    await test.step(`"Email" input should be visible`, async () => {
      await expect(this.emailInput).toBeVisible();
    });
  }

  async fillEmailInput(email: string) {
    await test.step(`Fill "Email" input with "${email}"`, async () => {
      await this.emailInput.fill(email);
    });
  }

  async setEmailInput(email: string) {
    await test.step(`Set "Email" input to "${email}"`, async () => {
      await this.emailInput.fill(email);
      await this.emailInput.blur();
    });
  }

  async emailErrorShouldBeVisible(errorMessage?: string) {
    if (!errorMessage) {
      await test.step(`"Email" error should be visible`, async () => {
        await expect(this.emailError).toBeVisible();
      });
    } else {
      await test.step(`"Email" error: "${errorMessage}"  should be visible`, async () => {
        await expect(this.emailError).toBeVisible();
        await expect(this.emailError).toHaveText(errorMessage);
      });
    }
  }

  async emailErrorShouldNotBeVisible() {
    await test.step(`"Email" error should not be visible`, async () => {
      await expect(this.emailError).not.toBeVisible();
    });
  }

  async passwordInputShouldBeVisible() {
    await test.step(`"Password" input should be visible`, async () => {
      await expect(this.passwordInput).toBeVisible();
    });
  }

  async fillPasswordInput(password: string) {
    await test.step(`Fill "Password" input with "${password}"`, async () => {
      await this.passwordInput.fill(password);
    });
  }

  async setPasswordInput(password: string) {
    await test.step(`Set "Password" input to "${password}"`, async () => {
      await this.passwordInput.fill(password);
      await this.passwordInput.blur();
    });
  }

  async passwordErrorShouldBeVisible(errorMessage?: string) {
    if (!errorMessage) {
      await test.step(`"Password" error should be visible`, async () => {
        await expect(this.passwordError).toBeVisible();
      });
    } else {
      await test.step(`"Password" error: "${errorMessage}"  should be visible`, async () => {
        await expect(this.passwordError).toBeVisible();
        await expect(this.passwordError).toHaveText(errorMessage);
      });
    }
  }

  async passwordErrorShouldNotBeVisible() {
    await test.step(`"Password" error should not be visible`, async () => {
      await expect(this.passwordError).not.toBeVisible();
    });
  }

  async signUpButtonShouldBeVisible() {
    await test.step(`"Sign up" button should be visible`, async () => {
      await expect(this.signUpButton).toBeVisible();
      await expect(this.signUpButton).toHaveText('Sign up');
    });
  }

  async signUpButtonShouldBeDisabled() {
    await test.step(`"Sign up" button should be disabled`, async () => {
      await expect(this.signUpButton).toBeDisabled();
      await expect(this.signUpButton).toHaveClass(/btn-disabled/);
    });
  }

  async signUpButtonShouldBeEnabled() {
    await test.step(`"Sign up" button should be enabled`, async () => {
      await expect(this.signUpButton).toBeEnabled();
      await expect(this.signUpButton).not.toHaveClass(/btn-disabled/);
    });
  }

  async clickSignUpButton() {
    await test.step('Click "Sign up" button', async () => {
      await this.signUpButton.click();
    });
  }

  async goSignInButtonShouldBeVisible() {
    await test.step(`"You already have an account?" button should be visible`, async () => {
      await expect(this.goSignInButton).toBeVisible();
      await expect(this.goSignInButton).toHaveText('You already have an account?');
    });
  }

  async clickGoSignInButton() {
    await test.step('Click "You already have an account?"', async () => {
      await this.goSignInButton.click();
    });
  }
}

File: tests/selenide/src/test/java/objects/forms/SignUpForm.java
package objects.forms;

import com.codeborne.selenide.SelenideElement;
import io.qameta.allure.Step;

import static com.codeborne.selenide.Condition.*;
import static utils.selectors.ElementSelectors.byTestId;

public class SignUpForm {

    private final SelenideElement container;
    private final SelenideElement userNameInput;
    private final SelenideElement userNameError;
    private final SelenideElement emailInput;
    private final SelenideElement emailError;
    private final SelenideElement passwordInput;
    private final SelenideElement passwordError;
    private final SelenideElement signUpError;
    private final SelenideElement signUpButton;
    private final SelenideElement goSignInButton;

    public SignUpForm(SelenideElement container) {
        this.container = container;
        this.userNameInput = container.$(byTestId("input-username"));
        this.userNameError = container.$(byTestId("error-username"));
        this.emailInput = container.$(byTestId("input-email"));
        this.emailError = container.$(byTestId("error-email"));
        this.passwordInput = container.$(byTestId("input-password"));
        this.passwordError = container.$(byTestId("error-password"));
        this.signUpError = container.$(byTestId("error-signup"));
        this.signUpButton = container.$(byTestId("button-signup"));
        this.goSignInButton = container.$(byTestId("link-go-signin"));
    }

    @Step("'Sign up' form should be visible")
    public SignUpForm shouldBeVisible() {
        this.userNameInputShouldBeVisible();
        this.emailInputShouldBeVisible();
        this.passwordInputShouldBeVisible();
        this.signUpButtonShouldBeVisible();
        this.goSignInButtonShouldBeVisible();
        return this;
    }

    @Step("Fill form with '{userName}', '{email}' and '{password}'")
    public SignUpForm fillForm(String userName, String email, String password) {
        this.fillUserNameInput(userName);
        this.fillEmailInput(email);
        this.fillPasswordInput(password);
        return this;
    }

    @Step("Set form to '{userName}', '{email}' and '{password}'")
    public SignUpForm setForm(String userName, String email, String password) {
        this.setUserNameInput(userName);
        this.setEmailInput(email);
        this.setPasswordInput(password);
        return this;
    }

    @Step("Errors should not be visible")
    public SignUpForm errorsShouldNotBeVisible() {
        this.userNameErrorShouldNotBeVisible();
        this.emailErrorShouldNotBeVisible();
        this.passwordErrorShouldNotBeVisible();
        return this;
    }

    @Step("'Username' input should be visible")
    public SignUpForm userNameInputShouldBeVisible() {
        this.userNameInput.shouldBe(visible);
        return this;
    }

    @Step("Fill 'Username' input with '{userName}'")
    public SignUpForm fillUserNameInput(String userName) {
        this.userNameInput.sendKeys(userName);
        return this;
    }

    @Step("Set 'Username' input to '{userName}'")
    public SignUpForm setUserNameInput(String userName) {
        this.userNameInput.setValue(userName);
        return this;
    }

    @Step("'Username' error should be visible")
    public SignUpForm userNameErrorShouldBeVisible() {
        this.userNameError.shouldBe(visible);
        return this;
    }

    @Step("'Username' error: '{errorMessage}' should be visible")
    public SignUpForm userNameErrorShouldBeVisible(String errorMessage) {
        this.userNameError.shouldHave(exactText(errorMessage));
        return this;
    }

    @Step("'Username' error should not be visible")
    public SignUpForm userNameErrorShouldNotBeVisible() {
        this.userNameError.shouldNotBe(visible);
        return this;
    }

    @Step("'Email' input should be visible")
    public SignUpForm emailInputShouldBeVisible() {
        this.emailInput.shouldBe(visible);
        return this;
    }

    @Step("Fill 'Email' input with '{email}'")
    public SignUpForm fillEmailInput(String email) {
        this.emailInput.sendKeys(email);
        return this;
    }

    @Step("Set 'Email' input to '{email}'")
    public SignUpForm setEmailInput(String email) {
        this.emailInput.setValue(email);
        return this;
    }

    @Step("'Email' error should be visible")
    public SignUpForm emailErrorShouldBeVisible() {
        this.emailError.shouldBe(visible);
        return this;
    }

    @Step("'Email' error: '{errorMessage}' should be visible")
    public SignUpForm emailErrorShouldBeVisible(String errorMessage) {
        this.emailError.shouldHave(exactText(errorMessage));
        return this;
    }

    @Step("'Email' error should not be visible")
    public SignUpForm emailErrorShouldNotBeVisible() {
        this.emailError.shouldNotBe(visible);
        return this;
    }

    @Step("'Password' input should be visible")
    public SignUpForm passwordInputShouldBeVisible() {
        this.passwordInput.shouldBe(visible);
        return this;
    }

    @Step("Fill 'Password' input with '{password}'")
    public SignUpForm fillPasswordInput(String password) {
        this.passwordInput.sendKeys(password);
        return this;
    }

    @Step("Set 'Password' input to '{password}'")
    public SignUpForm setPasswordInput(String password) {
        this.passwordInput.setValue(password);
        return this;
    }

    @Step("'Password' error should be visible")
    public SignUpForm passwordErrorShouldBeVisible() {
        this.passwordError.shouldBe(visible);
        return this;
    }

    @Step("'Password' error: '{errorMessage}' should be visible")
    public SignUpForm passwordErrorShouldBeVisible(String errorMessage) {
        this.passwordError.shouldHave(exactText(errorMessage));
        return this;
    }

    @Step("'Password' error should not be visible")
    public SignUpForm passwordErrorShouldNotBeVisible() {
        this.passwordError.shouldNotBe(visible);
        return this;
    }

    @Step("'Sign up' error should be visible")
    public SignUpForm signUpErrorShouldBeVisible() {
        this.signUpError.shouldBe(visible);
        return this;
    }

    @Step("'Sign up' error: '{errorMessage}' should be visible")
    public SignUpForm signUpErrorShouldBeVisible(String errorMessage) {
        this.signUpError.shouldHave(exactText(errorMessage));
        return this;
    }

    @Step("'Sign up' error should not be visible")
    public SignUpForm signUpErrorShouldNotBeVisible() {
        this.signUpError.shouldNotBe(visible);
        return this;
    }

    @Step("'Sign up' button should be visible")
    public SignUpForm signUpButtonShouldBeVisible() {
        this.signUpButton.shouldHave(exactText("Sign up"));
        return this;
    }

    @Step("'Sign up' button should be disabled")
    public SignUpForm signUpButtonShouldBeDisabled() {
        this.signUpButton.shouldBe(disabled);
        return this;
    }

    @Step("'Sign up' button should be enabled")
    public SignUpForm signUpButtonShouldBeEnabled() {
        this.signUpButton.shouldBe(enabled);
        return this;
    }

    @Step("Click 'Sign up' button")
    public SignUpForm clickSignUpButton() {
        this.signUpButton.click();
        return this;
    }

    @Step("'Go to sign in' button should be visible")
    public SignUpForm goSignInButtonShouldBeVisible() {
        this.goSignInButton.shouldHave(exactText("You already have an account?"));
        return this;
    }

    @Step("Click 'Go to sign in' button")
    public SignUpForm clickGoSignInButton() {
        this.goSignInButton.click();
        return this;
    }

}